source/MaterialXGenShader/Shader.h

```diff
@@ -31,291 +23,77 @@ using ShaderPtr = shared_ptr<class Shader>;
+    {
+        auto it = _attributeMap.find(attrib);
+        return it != _attributeMap.end() ? it->second : nullptr;
+    }
 
-    /// Return the active shader graph.
-    ShaderGraph* getGraph() const { return _graphStack.back(); }
+    /// Set a value attribute on the shader.
+    void setAttribute(const string& attrib, ValuePtr value)
+    {
+        _attributeMap[attrib] = value;
+    }
 
-    /// Push a new active shader graph.
-    /// Used when emitting code for compounds / subgraphs.
-    void pushActiveGraph(ShaderGraph* graph) { _graphStack.push_back(graph); }
+    /// Set a flag attribute on the shader.
+    void setAttribute(const string& attrib)
+    {
+        _attributeMap[attrib] = Value::createValue<bool>(true);
+    }
 
-    /// Reactivate the previously last used shader graph.
-    void popActiveGraph() { _graphStack.pop_back(); }
+    /// Return the shader graph.
+    const ShaderGraph& getGraph() const { return *_graph; }
 
     /// Return true if this shader matches the given classification.
-    bool hasClassification(unsigned int c) const { return getGraph()->hasClassification(c); }
+    bool hasClassification(unsigned int c) const { return _graph->hasClassification(c); }
 
     /// Return the final shader source code for a given shader stage
-    const string& getSourceCode(const string& stage = PIXEL_STAGE) const { return getStage(stage)->code; }
-
-protected:
-
-    /// A shader stage, containing the state and 
-    /// resulting source code for the stage
-    struct Stage
-    {
-        string name;
-        int indentations;
-        std::queue<Brackets> scopes;
-        std::set<string> includes;
-        std::set<ShaderNodeImpl*> definedFunctions;
-
-        // Block holding constant variables for this stage
-        VariableBlock constants;
-
-        // Blocks holding uniform variables for this stage
-        VariableBlockMap uniforms;
-
-        // Resulting source code for this stage
-        string code;
-
-        Stage(const string& n) : name(n), indentations(0), constants("Constants", "cn") {}
-    };
+    const string& getSourceCode(const string& stage = MAIN_STAGE) const { return getStage(stage).getSourceCode(); }
 
-    using StagePtr = std::shared_ptr<Stage>;
-
-    /// Create a new stage with given name.
-    StagePtr createStage(const string& name);
-
-    /// Return the stage with the given name.
-    Stage* getStage(const string& name);
-    const Stage* getStage(const string& name) const;
-
-    /// Add indentation on current line
-    virtual void indent();
-
-    /// Return a container with all top level graphs use by this shader.
-    virtual void getTopLevelShaderGraphs(ShaderGenerator& shadergen, std::deque<ShaderGraph*>& graphs) const;
+  protected: 
+    /// Create a new stage in the shader.
+    ShaderStagePtr createStage(const string& name, ConstSyntaxPtr syntax);
 
     string _name;
-    ShaderGraphPtr _rootGraph;
-    vector<ShaderGraph*> _graphStack;
-
-    Stage* _activeStage;
```

Comment:
Removed the concept of an active shader stage for editing. All edits are now done on an explicit ShaderStage instance, through the ShaderGenerator class.